contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1422 @@
+    function topUp(address operator, uint96 amount)
+        external
+        override
+        onlyOwnerOrOperator(operator)
+    {
+        require(amount > 0, "Amount must be greater than 0");
+        OperatorInfo storage operatorStruct = operators[operator];
+        operatorStruct.tStake += amount;
+        emit ToppedUp(operator, amount);
+        token.safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    /// @notice Propagates information about stake top-up from the legacy KEEP
+    ///         staking contract to T staking contract. Can be called only by
+    ///         the owner or operator.
+    function topUpKeep(address operator)
+        external
+        override
+        onlyOwnerOrOperator(operator)
+    {
+        OperatorInfo storage operatorStruct = operators[operator];
+        uint96 tAmount = getKeepAmountInT(operator);
+        require(tAmount > operatorStruct.keepInTStake, "Nothing to top-up");
+
+        emit ToppedUp(operator, tAmount - operatorStruct.keepInTStake);
+        operatorStruct.keepInTStake = tAmount;
+    }
+
+    /// @notice Propagates information about stake top-up from the legacy NU
+    ///         staking contract to T staking contract. Can be called only by
+    ///         the owner or operator.
+    function topUpNu(address operator)
+        external
+        override
+        onlyOwnerOrOperator(operator)
+    {
+        OperatorInfo storage operatorStruct = operators[operator];
+        uint96 tAmount = getNuAmountInT(operatorStruct.owner, operator);
+        require(tAmount > operatorStruct.nuInTStake, "Nothing to top-up");
+
+        emit ToppedUp(operator, tAmount - operatorStruct.nuInTStake);
+        operatorStruct.nuInTStake = tAmount;
+    }
+
+    //
+    //
+    // Undelegating a stake (unstaking)
+    //
+    //
+
+    /// @notice Reduces the liquid T stake amount by the provided amount and
+    ///         withdraws T to the owner. Reverts if there is at least one
+    ///         authorization higher than the sum of the legacy stake and
+    ///         remaining liquid T stake or if the unstake amount is higher than
+    ///         the liquid T stake amount. Can be called only by the owner or
+    ///         operator.
+    function unstakeT(address operator, uint96 amount)
+        external
+        override
+        onlyOwnerOrOperator(operator)
+    {
+        OperatorInfo storage operatorStruct = operators[operator];
+        require(
+            amount > 0 &&
+                amount + getMinStaked(operator, StakeType.T) <=
+                operatorStruct.tStake,
+            "Too much to unstake"
+        );
+        operatorStruct.tStake -= amount;
+        require(
+            operatorStruct.tStake >= minTStakeAmount ||
+                operatorStruct.startTStakingTimestamp + MIN_STAKE_TIME <=
+                /* solhint-disable-next-line not-rely-on-time */
+                block.timestamp,
+            "Can't unstake earlier than 24h"
```

Comment:
It reverts only if less than minimum and time earlier than 24h, so it really checks only time. But I'm fine to change

Reply:
OK, I think I was confused and this revert message makes sense.

We do not require a minimum T stake to be maintained all the time. After 24 hours, T stake can be reduced below the minimum stake. Minimum stake is just a measure of protection against griefing.

What if we add exactly this ☝🏻 explanation to `setMinimumStakeAmount` docs?

Reply:
Also, shouldn't `unstakeAll` have the same rule instead of `minTStakeAmount == 0 ||` ?

Reply:
> Also, shouldn't `unstakeAll` have the same rule instead of `minTStakeAmount == 0 ||` ?

`operatorStruct.tStake` will always be 0 after `unstakeAll` which means less than `minTStakeAmount` except `minTStakeAmount == 0`

Reply:
👌🏻 , let's leave it as-is and just explain what min stake is for future-us and auditors:
```
/// @notice Allows the Governance to set the minimum required stake amount.
///         This amount is required to protect against griefing the staking
///         contract and individual applications are allowed to require
///         higher minimum stakes if necessary.
/// @dev Operators are not required to maintain a minimum T stake all
///      the time. 24 hours after the delegation, T stake can be reduced
///      below the minimum stake.
function setMinimumStakeAmount(uint96 amount)
```

Reply:
✔️